File: javascript/xss_vulnerable.js
/**
 * Vulnerable JavaScript code with XSS flaws.
 * For security scanner testing only.
 */

function displaySearchResults(query) {
    // DOM-based XSS via innerHTML
    const resultsDiv = document.getElementById('results');
    resultsDiv.innerHTML = `<h1>Results for: ${query}</h1>`;
}

function greetUser(username) {
    // DOM-based XSS via innerHTML
    const greeting = document.getElementById('greeting');
    greeting.innerHTML = "<h1>Welcome " + username + "</h1>";
}

function showMessage(message) {
    // DOM-based XSS via outerHTML
    const messageDiv = document.getElementById('message');
    messageDiv.outerHTML = `<div class="alert">${message}</div>`;
}

function renderContent(content) {
    // DOM-based XSS via document.write
    document.write(content);
}

function updatePage(html) {
    // DOM-based XSS via document.writeln
    document.writeln(html);
}

function displayUserData(data) {
    // jQuery XSS
    $('#user-data').html(data);
}

function reactDangerousComponent(userContent) {
    // React dangerouslySetInnerHTML
    return <div dangerouslySetInnerHTML={{__html: userContent}} />;
}

function displaySearchResultsSafe(query) {
    // Safe version using textContent
    const resultsDiv = document.getElementById('results');
    resultsDiv.textContent = `Results for: ${query}`;
}

function greetUserSafe(username) {
    // Safe version using textContent
    const greeting = document.getElementById('greeting');
    greeting.textContent = `Welcome ${username}`;
}

File: python/xss_vulnerable.py
"""
Vulnerable Python code with XSS flaws.
For security scanner testing only.
"""
from flask import Flask, request, render_template_string
from django.utils.safestring import mark_safe

app = Flask(__name__)


@app.route('/search')
def search_vulnerable():
    """XSS via render_template_string with user input."""
    query = request.args.get('q', '')
    template = f"<h1>Results for: {query}</h1>"
    return render_template_string(template)


@app.route('/greet')
def greet_vulnerable():
    """XSS via string concatenation in template."""
    name = request.args.get('name', '')
    html = "<h1>Hello " + name + "</h1>"
    return render_template_string(html)


@app.route('/message')
def message_vulnerable():
    """XSS via format string in template."""
    msg = request.args.get('msg', '')
    return render_template_string("<div>{}</div>".format(msg))


def django_mark_safe_vulnerable(user_input):
    """Django XSS via mark_safe."""
    content = mark_safe(user_input)
    return content


@app.route('/profile')
def profile_safe():
    """Safe version using Jinja2 auto-escaping."""
    username = request.args.get('username', '')
    return render_template('profile.html', username=username)
